core/src/main/java/com/linecorp/armeria/internal/annotation/AnnotatedHttpServiceFactory.java

```diff
@@ -215,6 +223,86 @@
+        }
+        return Optional.of(repeatable.get().value());
+    }
+
+    @SuppressWarnings("unchecked")
+    private static <T extends Annotation> List<T> findRepeatableAnnotations(AnnotatedElement element,
+                                                                            Class<T> annotationType) {
+        requireNonNull(element, "element");
+        requireNonNull(annotationType, "annotationType");
+        final Class<? extends Annotation> repeatable = getRepeatable(annotationType).orElseThrow(
+                () -> new IllegalArgumentException("annotationType is not an repeatable annotation."));
+        return getAllAnnotations(element).stream()
+                                         .filter(a -> a.annotationType() == annotationType ||
+                                                      a.annotationType() == repeatable)
+                                         .flatMap(annotation -> {
+                                             if (annotation.annotationType() == repeatable) {
+                                                 try {
+                                                     return Arrays.stream((T[]) annotation.annotationType()
+                                                                                          .getMethod("value")
+                                                                                          .invoke(annotation));
+                                                 } catch (Exception ex) {
+                                                     throw new IllegalStateException(
+                                                             "value() method invocation throws an exception.",
+                                                             ex);
+                                                 }
+                                             }
+                                             return Stream.of((T) annotation);
+                                         }).collect(toImmutableList());
+    }
+
+    private static void setAdditionalHeader(HttpHeaders headers, AnnotatedElement element, String level) {
+        requireNonNull(headers, "headers");
+        requireNonNull(element, "element");
+        requireNonNull(level, "level");
+
+        final Set<String> addedHeaderSets = new HashSet<>();
+        findRepeatableAnnotations(element, AdditionalHeader.class).forEach(header -> {
+            if (addedHeaderSets.contains(header.name())) {
+                logger.warn("The additional header named '{}' is already added at the same {} level." +
+                            " It will be ignored.",
+                            header.name(), level);
+                return;
+            }
+            headers.set(HttpHeaderNames.of(header.name()), header.value());
+            addedHeaderSets.add(header.name());
+        });
+    }
+
+    private static void setAdditionalTrailer(HttpHeaders headers, AnnotatedElement element, String level) {
```

Comment:
Perhaps we can combine these two methods into one adding a type parameter which indicates whether it is `AdditionalHeader.class` or `AdditionalTrailer.class` so that we can reduce the duplicated code?

Reply:
e.g.

```java
private static void setAdditionalHeader(Class<?> annotationType, HttpHeaders headers, AnnotatedElement element, String level) {
    ...
    findRepeatableAnnotations(element, annotationType).forEach(header -> {
    });
    ...
}
```